app/src/main/java/org/schabi/newpipe/util/NavigationHelper.java

```diff
@@ -138,6 +155,231 @@ public static void playOnBackgroundPlayer(final Context context, final PlayQueue
         startService(context, getPlayerIntent(context, BackgroundPlayer.class, queue, resumePlayback));
     }
 
+    public static void downloadPlaylist(final PlaylistFragment context, final PlayQueue queue) {
+        List<PlayQueueItem> events = queue.getStreams();
+        Iterator<PlayQueueItem> eventsIterator = events.listIterator();
+        if (eventsIterator.hasNext()) {
+            startDownloadPlaylist(context, eventsIterator, null);
+        }
+    }
+
+    private static void startDownloadPlaylist(final PlaylistFragment activity,
+                                              final Iterator<PlayQueueItem> itemIterator,
+                                              DownloadSetting downloadSetting) {
+        if (downloadSetting != null) {
+            Completable.create(emitter -> {
+                while (itemIterator.hasNext()) {
+                    PlayQueueItem queueItem = itemIterator.next();
+                    queueItem.getStream().filter(streamInfo ->
+                            streamInfo != null).subscribe(streamInfo -> startDownloadFromDownloadSetting(activity,
+                            downloadSetting, streamInfo), activity::onError);
+                }
+                emitter.onComplete();
+            }).subscribe();
+        } else {
+            try {
+                if (itemIterator.hasNext()) {
+                    PlayQueueItem item = itemIterator.next();
+                    item.getStream().subscribe(streamInfo -> startDownloadFromStreamInfo(activity, streamInfo, itemIterator),
+                            activity::onError);
+                }
+            } catch (Exception e) {
+                Toast.makeText(activity.getActivity(),
+                        R.string.could_not_setup_download_menu,
+                        Toast.LENGTH_LONG).show();
+                e.printStackTrace();
+            }
+        }
+    }
+
+    /**
+     * Starts downloading video without invoking the {@link DownloadDialog}
+     *
+     * @param activity
+     * @param downloadSetting
+     * @param streamInfo
+     */
+    private static void startDownloadFromDownloadSetting(PlaylistFragment activity, DownloadSetting downloadSetting, StreamInfo streamInfo) {
+
+        DownloadSetting downloadSettingNew = refactorDownloadSetting(activity, downloadSetting, streamInfo);
+
+        if (downloadSettingNew == null) return;
+
+        DownloadManagerService.startMission(activity.getContext(), downloadSettingNew);
+    }
+
+    /**
+     * Refactors the download setting for the new stream info
+     *
+     * @param activity
+     * @param downloadSetting
+     * @param streamInfo
+     * @return
+     */
+    private static DownloadSetting refactorDownloadSetting(PlaylistFragment activity, DownloadSetting downloadSetting,
```

Comment:
This entire method just feels out of place, navigation should be the only thing done in this class.